Added support for node and npm based projects

davidfowl opened the pull request:
- Added AddNodeApp and AddNpmApp extension methods
- Flow the project's working directory into the host in 2 ways, via an attribute and as another project in the list. By default, executable's working directory are relative to the AppHost's project directory (this is what the attribute is used for).
- Added support for flowing the dynamically chosen port to the application by specifying a "portEnvVar" setting on the ServiceBindingAnnotation.
- Added `IResourceWithServiceDiscovery` which allows custom resources to export service discovery information.
- This test relies on node being installed.
- ~Added node.v0 resource for node apps~
- Improve support for executable resources in the manifest.

TODO: 
1. ~Test for npm based project.~
2. Skip tests if node isn't on the path.
3. ~Figure out what to do for deployment? Should we model this as a `node.v0` resource? That would match what we do for .NET projects...~

Fixes https://github.com/dotnet/aspire/issues/1017

PS: Fixing some of these issues (flowing port to the app and fixing working directories makes it easier to execute other project types in general). AddNpmApp and AddNodeApp are very thin wrappers now.

davidfowl committed:
Added suppot for node and npm based projects
- Added AddNodeApp and AddNpmApp extension methods
- Flow the project's working directory into the host in 2 ways, via an attribute and as another project in the list. By default, executable's working directory are relative to the AppHost's project directory (this is what the attribute is used for).
- Added support for flowing the dynamically chosen port to the application by specifying a "portEnvVar" setting on the ServiceBindingAnnotation.
- This test relies on node being installed.

davidfowl committed:
Added npm test

davidfowl committed:
Added newline

davidfowl committed:
Add WithReference overload for NodeAppResource

davidfowl commented on src/Aspire.Hosting/Node/NodeExtensions.cs:
```diff
@@ -0,0 +1,70 @@
+    /// <param name="builder">The resource where the service discovery information will be injected.</param>
+    /// <param name="source">The resource from which to extract service bindings.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{TDestination}"/>.</returns>
+    public static IResourceBuilder<TDestination> WithReference<TDestination, TSource>(this IResourceBuilder<TDestination> builder, IResourceBuilder<TSource> source)
```
@mitchdenny Now might be the time to fix this. Right now, this requires adding explicit overloads for things that support service discovery. We need to make it more contract based.

davidfowl committed:
Added IResourceWithServiceDiscovery
- This allows custom resources to opt-in to service discovery exports.
- Removed extra generic argument from WithReference

davidfowl commented on src/Aspire.Hosting/Node/NodeExtensions.cs:
```diff
@@ -0,0 +1,54 @@
+    private static IResourceBuilder<NodeAppResource> WithNodeDefaults(this IResourceBuilder<NodeAppResource> builder) =>
+        builder.WithOtlpExporter()
+            .WithEnvironment("NODE_ENV", builder.ApplicationBuilder.Environment.IsDevelopment() ? "development" : "production")
+            .ExcludeFromManifest();
```
This feels bad to bake into the method.

davidfowl committed:
Remove ExcludeFromManifest

davidfowl committed:
Fixed description

mitchdenny commented on src/Aspire.Hosting/ApplicationModel/ServiceBindingAnnotation.cs:
```diff
@@ -15,7 +15,7 @@ namespace Aspire.Hosting.ApplicationModel;
 public sealed class ServiceBindingAnnotation : IResourceAnnotation
 {
-    public ServiceBindingAnnotation(ProtocolType protocol, string? uriScheme = null, string? transport = null, string? name = null, int? port = null, int? containerPort = null, bool? isExternal = null)
+    public ServiceBindingAnnotation(ProtocolType protocol, string? uriScheme = null, string? transport = null, string? name = null, int? port = null, int? containerPort = null, bool? isExternal = null, string? portEnvVar = null)
```
Could ```portEnvVar``` just be ```env``` or ```environmentVariable```?

mitchdenny commented:
Overall this looks pretty solid.

davidfowl commented:
> I think if we go down the path of emitting a node.v0 resource that we'll have to give some kind of guidance on what we expect tools to do with it. For example, if azd or aspirate encounter a node.v0 resource should then look for a Dockerfile in the same path as package.json? Would there be an explciit override in package.json to specify the location of the Dockerfile?

Right, I think this is what we need to decide. Today for project.v0 resources (maybe this should be a dotnet.v0 resource) we recommend people use container builds, but that's all dependent on the target environment. You could imagine it meaning do a dotnet publish and zip the results and upload it to the target (It could mean use the nodejs build pack... etc)

I think docker file support needs to be an explicit call.

mitchdenny commented:
Regarding renaming `project.v0` to `dotnet.v0` ... I **think** this might make sense. Although we'd have to consider how things like `*.esproj` files might be handled if we want to explicitly support them. Maybe AddProject interrogates the project metadata and calls AddDotnet or AddNpm as appropriate.

What I mean about Docker support though is that I think as a framework orientated towards cloud native development we assume that if you have a reosurce like Node or Python apps that there will be a Docker build involved in deployment.

davidfowl committed:
PR feedback and support for manifest
- Renamed PortEnvVar to env
- Added manifest resource for node apps (node.v0)
- Cleaned up executable support

mitchdenny reviewed (approved):
Assuming ```WriteNodeApp``` is temporarily in ```ManifestPublisher``` until the ```ManifestPublisherCallbackAnnotation``` is updated to take a context.

davidfowl committed:
Remove the node.v0 resource type

davidfowl committed:
Fix the relative path for working directory generation

davidfowl committed:
Skip node if its not installed

davidfowl commented on src/Aspire.Hosting/build/Aspire.Hosting.targets:
```diff
@@ -63,13 +63,51 @@ public class ]]>%(ClassName)<![CDATA[ : IServiceMetadata
     </ItemGroup>
   </Target>
 
+  <Target Name="CreateProjectMetadata">
```
@baronfel can you give me an after the fact review on this?

baronfel replied:
This all seems ok to me, seems straightforward enough based on the existing patterns.
